name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/MerkurPrivatBankPDFExtractor.java

```diff
@@ -0,0 +1,124 @@
+                        .match("^(?<name1>.*)$")
+                        .match("^Ausf.hrungskurs [\\.,\\d]+ (?<currency>[\\w]{3}) .*$")
+                        .assign((t, v) -> {
+                            if (!v.get("name1").startsWith("Handels-/Ausführungsplatz"))
+                                v.put("name", trim(v.get("name")) + " " + trim(v.get("name1")));
+                            t.setSecurity(getOrCreateSecurity(v));
+                        })
+                        
+                        // @formatter:off
+                        // Stück 125,3258 XTR.(IE) - MSCI WORLD              IE00BJ0KDQ92 (A1XB5U)
+                        // @formatter:on
+                        .section("shares")
+                        .match("^St.ck (?<shares>[\\.,\\d]+) .*$")
+                        .assign((t, v) -> t.setShares(asShares(v.get("shares"))))
+                        
+                        // @formatter:off
+                        // Schlusstag/-Zeit 02.05.2023 09:34:40 Auftraggeber Max Mustermann
+                        // @formatter:on
+                        .section("date", "time")
+                        .match("^Schlusstag\\/\\-Zeit (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) (?<time>[\\d]{2}:[\\d]{2}:[\\d]{2}) .*$")
+                        .assign((t, v) -> t.setDate(asDate(v.get("date"), v.get("time"))))
+
+                        // @formatter:off
+                        // Ausmachender Betrag 10.002,50- EUR
+                        // @formatter:on
+                        .section("amount", "currency")
+                        .match("^Ausmachender Betrag (?<amount>[\\.,\\d]+)\\- (?<currency>[\\w]{3})$")
+                        .assign((t, v) -> {
+                            t.setAmount(asAmount(v.get("amount")));
+                            t.setCurrencyCode(asCurrencyCode(v.get("currency")));
+                        })
+                        
+                        // @formatter:off
+                        // Auftragsnummer 284722/61.00
+                        // Ausf. erfolgte über Quotrix
+                        // Ihr ETF-Sparplan Nr.      1
+                        // Für das Geschäft wurde keine Anlageberatung erbracht.
+                        // @formatter:on
+                        .section("note", "note1").optional()
+                        .match("^ (?<note>Auftragsnummer .*)$")
+                        .match("^Ausf. erfolgte über .*$")
+                        .match("^(?<note1>.*)$")
+                        .match("^Für das Gesch.ft wurde keine Anlageberatung erbracht.")
+                        .assign((t, v) -> {
+                            if (!v.get("note1").isEmpty())
+                                v.put("note", trim(v.get("note")) + "\n" + trim(replaceMultipleBlanks(v.get("note1"))));
+                            t.setNote(trim(v.get("note")));
+                        })
```

Comment:
The last change, then I think the PDF importer stands solidly there.
`.match("^ (?<note>Auftragsnummer .*)$")` <-- here is a blank

Here you have inserted a space. But there is a possibility that the formatting changes.
Change this blank to `.*`

`note1` can never be empty, because there must be a "match" here.
So the correct syntax is:
```
.section("note", "note1").optional()
	.match("^.*(?<note>Auftragsnummer .*)$")
	.find("Ausf\\. erfolgte .ber .*")
	.match("^(?<note1>.*)$")
	.match("^F.r das Gesch.ft .*$")
	.assign((t, v) -> t.setNote(trim(v.get("note")) + "\n" + trim(replaceMultipleBlanks(v.get("note1")))))
})
```
The ".find" --> [look here](https://github.com/buchen/portfolio/blob/21334c1e4668b2809170f35c986b3023ae797bb2/name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/PDFParser.java#L470:L474) <--

... and escaping special charaters

Reply:
Awesome. Thanks!

- When would you recommend to use the .find over the .match?
- The auto formatter on save breaks the assign into two lines on save due to line length. I assume that is ok in terms of the project codestyle?

> Activate Format Source Code and then Format edited lines

and

> Format the source code. The formatter configuration is part of the project source code. Exception: Do not reformat the PDF importer source code. Instead, carefully insert new code into the existing formatting.

Reply:
Roughly speaking, `find` and `match` is the same thing.
the difference is in the function.

**find syntax** `Pattern.compile("^" + string + "$")`
The anchors are already included and we use this to find a line and do not use attributes.

**match syntax `Pattern.compile(regex)`
In `match` we look for a direct match according to the syntax we want and the attributes.

> The auto formatter on save breaks the assign into two lines on save due to line length. I assume that is ok in terms of the project codestyle?

Yes, thats fine.
The problem is that codestyle makes the importers very messy.
So that the individual sections are easy to read and better to maintain, they should not be formatted. 
Otherwise we would have to force a lot of line breaks (by `//`) and that makes the sourcecode unreadable again.